nornir/plugins/inventory/netbox.py

```diff
@@ -0,0 +1,75 @@
+import os
+from builtins import super
+
+from nornir.core.inventory import Inventory
+
+import requests
+
+
+class NBInventory(Inventory):
+
+    def __init__(
+        self,
+        nb_url=None,
+        nb_token=None,
+        use_slugs=True,
+        flatten_custom_fields=True,
+        **kwargs
+    ):
+
+        nb_url = nb_url or os.environ.get("NB_URL", "http://localhost:8080")
+        nb_token = nb_token or os.environ.get(
+            "NB_TOKEN", "0123456789abcdef0123456789abcdef01234567"
+        )
+        headers = {"Authorization": "Token {}".format(nb_token)}
+
+        # Create dict of hosts using 'devices' from NetBox
+        devices = {}
+        nb_devices = requests.get(
+            "{}/api/dcim/devices/?limit=0".format(nb_url), headers=headers
+        ).json()
+
+        try:
+            for d in nb_devices["results"]:
+
+                # Create temporary dict
+                temp = {}
+
+                # Add value for IP address
+                if d.get("primary_ip", {}):
+                    temp["nornir_host"] = d["primary_ip"]["address"].split("/")[0]
+
+                # Add values that don't have an option for 'slug'
+                temp["serial"] = d["serial"]
+                temp["vendor"] = d["device_type"]["manufacturer"]["name"]
+                temp["asset_tag"] = d["asset_tag"]
+
+                if flatten_custom_fields:
+                    for cf, value in d["custom_fields"].items():
+                        temp[cf] = value
+                else:
+                    temp["custom_fields"] = d["custom_fields"]
+
+                # Add values that do have an option for 'slug'
+                if use_slugs:
+                    temp["site"] = d["site"]["slug"]
+                    temp["role"] = d["device_role"]["slug"]
+                    temp["model"] = d["device_type"]["slug"]
+
+                    # Attempt to add 'platform' based of value in 'slug'
+                    temp["nornir_nos"] = d["platform"]["slug"] if d["platform"] else None
+
+                else:
+                    temp["site"] = d["site"]["name"]
+                    temp["role"] = d["device_role"]
+                    temp["model"] = d["device_type"]
+                    temp["nornir_nos"] = d["platform"]
+
+                # Assign temporary dict to outer dict
+                devices[d["name"]] = temp
+
+        except KeyError:
```

Comment:
@bdlamprecht when this happens, what status code does nsot return? I am wondering if we can just do `raise_for_status`.

Reply:
It should return either a 401 or 403, otherwise, that's clearly a bug on `Netbox` and not very nice from an API perspective.

I think it's better if we test though. Would you mind changing:

```
nb_devices = requests.get(
      "{}/api/dcim/devices/?limit=0".format(nb_url), headers=headers
).json() 
```

with:

```
r = requests.get("{}/api/dcim/devices/?limit=0".format(nb_url), headers=headers)
r.raise_for_status()
nb_devices = r.json()
```

And test it with both a wrong and a correct token?

Reply:
With that change, it does appear to return a `403` with an invalid token:
```
HTTPError: 403 Client Error: Forbidden for url: http://localhost:8080/api/dcim/devices/?limit=0
```

Is that all you needed me to check on?

Reply:
Thanks! That's what I was hoping for. I will commit a couple of more changes and I think we will be ready to go here.

Reply:
Awesome! :100: 
I'm happy to test it out once you get it all cleaned up the way that you'd like.
Cheers